Add flag-evaluation spec, Makefile, initial spec structure (#35)
open-feature/spec: specification/flag-evaluation/flag-evaluation.md

>
> - name (optional): A logical string identifier for the client.

```
// example client creation and retrieval
OpenFeature.getClient({
  name: 'my-openfeature-client'
});
```

The name is a logical identifier for the client.

#### Client Usage

##### Requirement 1.6

> The client **MUST** provide a method to add `hooks` which accepts one or more API-conformant `hooks`, and appends them to the collection of any previously added hooks. When new hooks are added, previously added hooks are not removed.

```
// example hook attachment
client.addHooks([new MyHook()]);
```

See [hooks](./hooks.md) for details.

#### Flag Evaluation

##### Requirement 1.7

> The `client` **MUST** provide methods for flag evaluation, with parameters `flag key` (string, required), `default value` (boolean | number | string | structure, required), `evaluation context` (optional), and `evaluation options` (optional).

```
// example flag evaluation
var myValue = client.getValue('my-flag', false);
```

##### Condition 1.8

> The language type system differentiates between strings, numbers, booleans and structures.
>

> > ##### Conditional Requirement 1.8.1

Reviewer: I still don't think we want these in blockquotes.

Author: I agree, but for now our fancy little W3C-QA-compliant spec parser requires it. I will make an issue to resolve that. I agree it looks ugly.

Added chore issue: https://github.com/open-feature/spec/issues/37

>
> > The `client` **MUST** provide methods for typed flag evaluation, including boolean, numeric, string, and structure.
>
> > ```
> > // example boolean flag evaluation
> > boolean myBool =  client.getBooleanValue('bool-flag', false);
> >
> > // example overloaded string flag evaluation with optional params
> > string myString = client.getStringValue('string-flag', 'N/A', evaluationContext, options);
> >
> > // example number flag evaluation
> > number myNumber = client.getNumberValue('number-flag', 75);
> >
> > // example overloaded structure flag evaluation with optional params
> > MyStruct myStruct = client.getObjectValue<MyStruct>('structured-flag', { text: 'N/A', percentage: 75 }, evaluationContext, options);
> > ```
>
> > See [evaluation context](../evaluation-context/evaluation-context.md) for details.
>
> ##### Conditional Requirement 1.8.2
>

> > > The `client` **SHOULD** guarantee the returned value of any typed flag evaluation method is of the expected type. If the value returned by the underlying provider implementation does not match the expected type, it's to be considered abnormal execution, and the supplied `default value` should be returned.

Author: Yes, this is a good point. I could see how that would be an issue. We can say something about logging.

> > > The `client` **SHOULD** guarantee the returned value of any typed flag evaluation method is of the expected type. If the value returned by the underlying provider implementation does not match the expected type, it's to be considered abnormal execution, and the supplied `default value` should be returned.

Reviewer 2: That feels right to me. This is the requirement afterall, so just saying "it should log the error" is good enough for now

##### Requirement 1.9

> > The `client` **MUST** provide methods for detailed flag value evaluation with the value wrapped in a `evaluation details` structure, with parameters `flag key` (string, required), `default value` (boolean | number | string | structure, required), `evaluation context` (optional), and `evaluation options` (optional).

Author: These are the parameters to the method, not the fields in the structure. I will rephrase to make that more clear.

> > The `client` **MUST** provide methods for detailed flag value evaluation with the value wrapped in a `evaluation details` structure, with parameters `flag key` (string, required), `default value` (boolean | number | string | structure, required), `evaluation context` (optional), and `evaluation options` (optional).

Author: Again, these are the parameters to the method, not the fields in the structure. I will rephrase to make that more clear. The structure is detailed in `types.md`, as well as below.

> > The `client` **MUST** provide methods for detailed flag value evaluation with the value wrapped in a `evaluation details` structure, with parameters `flag key` (string, required), `default value` (boolean | number | string | structure, required), `evaluation context` (optional), and `evaluation options` (optional).

Author: rephrased, let me know if that's clearer.

> > The `client` **MUST** provide methods for detailed flag value evaluation with parameters `flag key` (string, required), `default value` (boolean | number | string | structure, required), `evaluation context` (optional), and `evaluation options` (optional), which returns an `evaluation details` structure.

Author: Basically. All the additional fields besides value are optional if you check the `.types.md` file.

> The language supports generics (or an equivalent feature).
>
> ##### Conditional Requirement 1.11.1
>
> > The `evaluation details` structure **SHOULD** accept a generic argument (or utilize an equivalent language feature) which indicates the type of the wrapped `value` field.

##### Requirement 1.12

> The `evaluation details` structure's `flag key` field **MUST** contain the `flag key` argument passed to the detailed flag evaluation method.

##### Requirement 1.13

> The `evaluation details` structure's `variant` field **SHOULD** contain a identifier logically corresponding to the return flag value.

For example, the flag value might be `3.14159265359`, and the variant field's value might be `"pi"`.

##### Requirement 1.14

> The `evaluation details` structure's `reason` field **SHOULD** indicate the semantic reason for the returned flag value. Possible values vary by provider, but may include `"TARGETING_MATCH"`, `"SPLIT"`, `"DISABLED"`, `"DEFAULT"`, `"UNKNOWN"` and `"ERROR"`.

##### Requirement 1.15

> In cases of abnormal execution (network failure, unhandled error, etc) the `reason` field in the `evaluation details` **SHOULD** be set to `"ERROR"`.

##### Requirement 1.16

> > In cases of abnormal execution, the `evaluation details` structure's `error code` field **SHOULD** identify an error occurred during flag evaluation, with possible values `"PROVIDER_NOT_READY"`, `"FLAG_NOT_FOUND"`, `"PARSE_ERROR"`, `"TYPE_MISMATCH"`, or `"GENERAL"`.

Reviewer 2: Is it worth stating as a requirement that in the case of normal execution, the `error code` field should not be set?

Author: Yes, I think that's a good point.

##### Requirement 1.17

> If the `reason` in the `evaluation details` is set to `"ERROR"`, the sibling `error code` field **MUST** be set.

##### Requirement 1.18

> The `evaluation options` structure's `hooks` field denotes a collection of hooks to be executed for the respective flag evaluation.

See [hooks](./hooks.md) for details.

##### Requirement 1.19

> No methods, functions, or operations on the client should ever throw exceptions, or otherwise abnormally terminate. Flag evaluation calls must always return the `default value` in the event of abnormal execution. Exceptions include functions or methods for the purposes for configuration or setup.

> ##### Requirement 1.20

Reviewer 3: I think we should take a harder stance here. Either the SDK supports async evaluation (and the Application Author has to plan for that) or the SDK does not support it and our spec should have explicitly defined semantics for what happens when you ask for a flag evaluation but it can't be done synchronously.

The specific scenario I'm thinking about is flagging for clientside SPAs,  where you often need to evaluate a flag at the point a page is loading, where they may well not be any locally cached flag configuration loaded yet.

Author: I think we need to consider languages and runtimes where asynchronous methods are not available or less common. For example, in many Java frameworks, thread-per-request is quite standard, so blocking while the flag is evaluated seems pretty reasonable there. Only a single request is blocked.

I see your concern, I just dont want to make it a hard requirement for all languages, especially ones in which blocking is common, or even idiomatic.

Reviewer 2: Why not say that the client **should** provide both async and sync mechanisms?
One consumer will probably want to almost-always call async methods, yet another consumer will probably want to almost-always call sync methods. Forcing only async to the consumer will likely result in just a lot of unnecessary `async` (JS) or `.get()' (Java future) calls, which they will probably write a simple wrapper around - so why not make that wrapper part of the spec?

It's pretty trivial for a provider to turn an async method into a sync one by using `await` `.get()` if their vendor-specific implementation only provides async methods, or turn a sync method into an async one by wrapping with `new Promise` etc.

Author: > Why not say that the client should provide both async and sync mechanisms?

I don't think this is reasonable, or possible in Javascript especially, given that we have vendors SDKs with ONLY async APIs already. If a method returns a promise, there's no way to make that method synchronous and also behave predictably without blocking the thread of execution (which means blocking your entire webapp in the case of node).

You're right that it's certainly possible to make a synchronous call async by wrapping it in a promise, but you cant go the other way. I also didn't want to force languages/frameworks where blocking is reasonable NOT to block (Java). Of course, I'm not against Futures or callbacks, but its certainly not unusual for Java calls to block.

Reviewer 2: yeah, you're right. I was thinking it is trivial to just `await` it, but of course, that makes your calling function async, then it's turtles all the way down.

so I wonder about saying "the client **should**" provide both a sync and an async method" (for those who want to call it in either form". If a vendor only provides an async method (and the language can't sync-ify it) then one option _could_ be to throw a `NotImplementedException` or similar. Though I do agree with the concept of 'our evaluation APIs shouldn't throw errors'. Hmmm.... I'll have a think. What are your thoughts in general to providing both sync and async evaluation methods?

Reviewer 2: Ooooh, thought.
one pattern used elsewhere would be `OpenFeature.getSyncClient(...)`  and `OpenFeature.getAsyncClient(...)`. That way, if needed, the getSyncClient could throw an error (legal in the current spec) without the `client.getStringValue()` itself needing to throw the exception to go "this isn't supported". Thoughts?

Author: > What are your thoughts in general to providing both sync and async evaluation methods?

I think it makes perfect sense in some implementation languages, but not others. It's perfectly reasonable for Java, but not JS. Like I mentioned before, it's pretty common for JS libs not to have any sync methods at all, and for good reason.

We could change it to a **SHOULD** that would just force the implementer to consider this question, not sure how to word it, but I'll think on it and respond.

Author: > Ooooh, thought.
one pattern used elsewhere would be OpenFeature.getSyncClient(...) and OpenFeature.getAsyncClient(...). That way, if needed, the getSyncClient could throw an error (legal in the current spec) without the client.getStringValue() itself needing to throw the exception to go "this isn't supported". Thoughts?

I think this is a reasonable idea, and might be a good path forward. I suppose my only concern is whether it's possible that there's languages where this concept itself wouldn't be idiomatic. I don't think Golang, with it lightweight goroutine concept and channels, typically sees explicitly async APIs. In Go, it's pretty easy to start a goroutine and wait for it to write to a channel without needing to "mark" the entire stack as async, like you would for async/await in c# or JS.

Reviewer: My Java implementation is blocking. 

I like the notion of a nonblocking version but I'm unsure if it should be the same repo or a different one. There are a *bunch* of methods we're already providing in the Java sdk. I'm not the most excited about doubling the count, but could do it. I'll be focusing on thread safety first.

Reviewer 2: @toddbaert @moredip , for the languages where it doesn't make sense to have one or other of the sync or async client/methods, we could just omit those `get[A]SyncClient(...)` methods from the API.

Author: OK, so to make sure I understand:

In implementation languages where there's both async and sync methods, we'd recommend (require?) the API to define 2 methods:

`openfeature.getAsyncClient()` and `openfeature.getSyncClient()` (these can abnormally terminate with `NotImplemented`, since there are "configuration")

And in languages where that dichotomy doesn't really exist, it's just:

`openfeature.getClient()`

If that's what you're suggesting, I think we can reflect that in the spec, and I think it's a pretty solid plan, so it has my vote. Let's wait to see if anyone objects, and if not, I'll modify the spec accordingly.

Let me know if I've misunderstood.

Reviewer 2: That's _very almost_ what I was suggesting 👍 .

And in languages where that dichotomy doesn't really exist, it's just:

`openfeature.getSyncClient()` or `openfeature.getAsyncClient()` (as appropriate)

IMO this keeps consistency across the different languages (and makes documentation easier for us) and is _ever so slightly_ more informative to the end-user... but TBH this slight modification is a tiny point and I'm happy with either option.

Reviewer 3: It feels pretty heavy to have two distinct clients (sync vs async), with two distinct interfaces, **AND** different semantics about how certain error scenarios are handled.

I suspect we're "only" really talking about this because of JS support, where you can't have a blocking call. I wonder if it would be better to add a special case for that scenario - and specify that "clients which cannot support a blocking method call should instead provide an equivalent asynchronous variant" or something like that?

Author: > AND different semantics about how certain error scenarios are handled.

I don't think different error semantics were being proposed, though maybe I missed something. There's just a stipulation that no errors are thrown with flag evaluation, but an exception to that rule is configuration/setup code, which includes getting a client. This means, to @agentgonzo 's point, that a `NotImplemented` could be thrown.

> I suspect we're "only" really talking about this because of JS support, where you can't have a blocking call. I wonder if it would be better to add a special case for that scenario - and specify that "clients which cannot support a blocking method call should instead provide an equivalent asynchronous variant" or something like that?

Yes, that, combined with the fact with already have a provider with an async underlying SDK, is the only reason I added anything about asynchrony in the spec at all, and perhaps that was a mistake. I think for now we should merge without having this nailed down, and we can revisit this specific issue later.

Reviewer: @toddbaert I'm not sure. I think we leave this as is, merge it, then discuss in an issue.

I would want to think through what a sync hook vs a sync hook might look like before I committed one way or the other, but this is good enough to make progress on imo.

# Flag Evaluation API

## Overview

> The `evaluation API` allows for the retrieval of feature flag values, independent of any flag control plane or vendor. In the absence of a feature flag system implementation (a [provider](../provider/providers.md)) the `evaluation API` uses the "No-op provider", which simply returns the supplied default flag value.

Reviewer 4: minor nit, should we rephrase this as `return of feature flag values` in place of `retrieval`. 

Is it worth calling out optional reason/metadata of why it was returned/error info?

Author: Thinking about it more, I think "evaluation" is probably the best word. What do you think?

I also agree it's worth mentioning the metadata, yes.

> The `evaluation API` allows for the retrieval of feature flag values, independent of any flag control plane or vendor. In the absence of a feature flag system implementation (a [provider](../provider/providers.md)) the `evaluation API` uses the "No-op provider", which simply returns the supplied default flag value.

Reviewer 5: I would either use the term "feature flag system implementation" or "provider" and not both. If you insist on using both I would put it the other way around as "provider" is the official name and "feature flag system implementation" is a description of what it is. IMO the link to the provider spec explains it sufficiently though.

#### Initialization and Configuration

##### Requirement 1.1

> The `API`, and any state it maintains **SHOULD** exist as a global singleton, even in cases wherein multiple versions of the `API` are present at runtime.

> It's important that multiple instances of the `API` not be active, so that state stored therein, such as the registered `provider`, static global `evaluation context`, and globally configured `hooks` allow the `API` to behave predictably. This can be difficult in some runtimes or languages, but implementors should make their best effort that only a single instance of the `API` is used.

Reviewer 4: I did not realize this was a consideration. It is not what we've seen in practice. It is generally how users work but not all of them. We have customers that instantiate multiple clients in an application.

Author: There's no problem with multiple clients (see 1.5). This is just a requirement for the API. A global singleton is necessary for advanced use cases, like "injecting" implementation into 3rd party libraries. If a third party implements the flag API, feature flagging can be used therein because it will be referencing the same global singleton, configured with a particular provider.

##### Requirement 1.2

> > The `API` **MUST** provide a function to set the `provider`, which accepts an API-conformant `provider` implementation.

Reviewer 5: ```suggestion
> The `API` **MUST** provide a function to set the global `provider` singleton, which accepts an API-conformant `provider` implementation.
```

```
// example provider mutator
OpenFeature.setProvider(new MyProvider());
```

see [provider](../provider//providers.md)

##### Requirement 1.3

> > The `API` **MUST** provide a function for `hook` addition, which accepts an array of API-conformant `hooks`. This adds the specified hooks to any hooks already attached.

Author: I've intentionally not gone into detail here, since `hooks` will have their own section.

Reviewer 5: Specifying it must be an array is an implementation detail IMO and depends on the idioms of the implementation language. In some languages it is idiomatic to have variadic arguments for example.

```suggestion
> The `API` **MUST** provide a function to register one or more `hooks`. This adds the specified hooks to any hooks already attached.
```

"This adds the specified hooks to any hooks already attached" reads weird to me. I sort of get what you're going for (that it doesn't replace existing hooks, but includes new ones) but it seems like a truism "addHooks adds hooks"

```suggestion
> The `API` **MUST** provide a function for `hook` addition, which accepts an array of API-conformant `hooks`. When new hooks are added, previously added hooks are not removed.
```

This also makes execution order ambiguous. I would either specify that it is not order guaranteed, or use a word like "append" or add a sentence like "hooks are executed in the order they are added".

```suggestion
> The `API` **MUST** provide a function for `hook` addition, which accepts an array of API-conformant `hooks`. This adds the specified hooks to any hooks already attached. Hooks are executed in the order they are added.
```

And all together:

```suggestion
> The `API` **MUST** provide a function to register one or more `hooks` which accepts an array of API-conformant `hooks`. When a new hook is added, previously added hooks are not removed. Hooks are executed in the order they are added.
```

Author: @dyladan  Good feedback here, I've amalgamated a few of these and updated.

We will have an entire hook section, and I wanted to talk about ordering in that document in some more detail. To that end, we'd discussed `before` hooks running in the order they are added, with all other hooks (`after`, `error`, `finally`, which run after evaluation) running in reverse order. Do you think that's too much to add here?

```
// example hook attachment
OpenFeature.addHooks([new MyHook()]);
```

see [hooks](./hooks.md)

##### Requirement 1.4

> The API **MUST** provide a function for retrieving the `provider` implementation.

```
// example provider accessor
OpenFeature.getProvider();
```

see [provider](../provider/provider.md)

##### Requirement 1.5

> > The `API` **MUST** provide a function for creating a `client`, with an options object containing an optional field for the client `name`.

Author: Is an "options object" too prescriptive? One one hand, it seems less likely to cause breaking changes in many languages, since new properties can be added, on the other hand, it might not be idiomatic in some cases.

Reviewer 4: if the `name` is optional, it does seem like the configuration itself should be an optional field. But if at a minimum it's required to pass a `name` then the options object makes sense being required.

Author: I'm OK with that. I think the only thing to consider is that in some languages it's less of a  breaking change to add new optional fields to an object, so things are slightly less brittle. I dont even know if that's true in Java though, I don't think so.

I will make this phrase less opinionated.

Author: > if the name is optional, it does seem like the configuration itself should be an optional field. But if at a minimum it's required to pass a name then the options object makes sense being required.

Ya, you're right, I guess this whole options field can be optional in languages that use an object.

Reviewer 5: The wording seems too oriented around the languages you're familiar with and similar to the above comment about the array of hooks seems to me an implementation detail which could have different idioms in different languages. For example optional arguments are often keyword arguments in python.

```suggestion
> The `API` **MUST** provide a function for creating a `client` which accepts the following options:
>
> - name (optional): A logical string identifier for the client.
```

Author: Yep, this resolves @justinabrahms point too.

```
// example client creation and retrieval
OpenFeature.getClient({
  name: 'my-openfeature-client'
});
```

The name is a logical identifier for the client.

> ##### Requirement 1.6

Reviewer 5: I was very confused by the sudden change from API to client. Maybe make a heading or even a new file?

Author: Agreed. I will add a heading. If this file starts to get too big, it might make sense to break it up, but I'm not convinced it's justified yet.

> > The client **MUST** provide a method for `hook` addition, which accepts an array of API-conformant `hooks`. This adds the specified hooks to any hooks already attached.

Author: I've intentionally not gone into detail here, since `hooks` will have their own section.

```
// example hook attachment

> client.addHooks([new MyHook()]);

Reviewer 5: API hooks _and_ client hooks? What is the difference? All my previous comments about that function apply to this one.

> client.addHooks([new MyHook()]);

Author: > API hooks and client hooks? What is the difference? All my previous comments about that function apply to this one.

There's only one API, but n number of clients. Clients might have a much smaller scope than the API, and attaching hooks to it only applies to flags evaluated using that client.

I was going to mention more about this in the `hooks` spec.

```

see [hooks](./hooks.md)

#### Flag Evaluation

##### Requirement 1.7

> > The `client` **MUST** provide methods for typed flag value evaluation, including boolean, numeric, string, and map (object), with parameters `flag key` (string, required), `default value` (boolean | number | string | object, required), `evaluation context` (optional), and `evaluation options` (optional).

Author: Is the `boolean | number | string | object` syntax clear?

Reviewer 5: Not at all clear to the uninitiated what "typed flag value evaluation" is

Reviewer 5: See my comments on addHooks about language agnosticism. Map/object may not apply to all languages you want to target, and many languages differentiate between Map and object.  I would name and define this concept, and let any language use whatever mechanism is appropriate. Is it a map from string keys to (number | string | boolean)? Can the object have nested objects? Can the object have non-string keys? Can the values be arrays? If so, must those arrays be homogeneous?

Author: I think it might be best to say something suggesting that the language use whatever construct is usually used to interrogate and manipulate JSON objects, that's really what these are in most backends.

@justinabrahms @dyladan what do you think?

Reviewer 5: JSON or arbitrary structured data? Could easily have a system which transmits these in YAML for example. Would that provider be expected to convert to a JSON specific structure?

```suggestion
> The `client` **MUST** provide methods for typed flag value evaluation, including boolean, numeric, string, and structured data, with parameters `flag key` (string, required), `default value` (boolean | number | string | object, required), `evaluation context` (optional), and `evaluation options` (optional). Structured data **SHOULD** be presented however is idiomatic in the implementation language for arbitrarily structured data such as JSON or YAML.
```

```
// example boolean flag evaluation
boolean myBool =  client.getBooleanValue('bool-flag', false);

// example overloaded string flag evaluation with optional params
string myString = client.getStringValue('string-flag', 'N/A', evaluationContext, options);

```

see [evaluation context](../evaluation-context/evaluation-context.md)

##### Requirement 1.8

> > The `client` **MUST** provide a method for flag activity evaluation, denoting if the flag is "enabled" or not, with parameters `flag key` (string, required), `default value` (boolean, required), `evaluation context` (optional), and `evaluation options` (optional).

Author: see: https://github.com/open-feature/spec/pull/35/files/a782d675f1aef6924092f7b9421a53c2f2ce449f#r855378994

> > The `client` **MUST** provide a method for flag activity evaluation, denoting if the flag is "enabled" or not, with parameters `flag key` (string, required), `default value` (boolean, required), `evaluation context` (optional), and `evaluation options` (optional).

Reviewer 5: "flag activity evaluation" means nothing to people who don't already know what it means. What does it mean for a flag to be enabled? Does this mean it has a value of true, or that it exists in the backend?

> > The `client` **MUST** provide a method for flag activity evaluation, denoting if the flag is "enabled" or not, with parameters `flag key` (string, required), `default value` (boolean, required), `evaluation context` (optional), and `evaluation options` (optional).

Author: You're right. I think we might be removing this completely based on the [this](https://github.com/open-feature/spec/pull/35#discussion_r855574278) discussion.

```
// example flag activity evaluation
boolean enabled = client.isEnabled('my-flag', false);
```

##### Requirement 1.9

> The `client` method for flag activity evaluation **MAY** be semantically different from the boolean flag evaluation method.

> Some providers may draw a distinction between boolean flag value evaluation, and flag activity. Put another way, a flag may be "disabled", but simultaneously have an associated auxillary boolean payload of `true`.

Reviewer 4: Is there additional information around this? I'm having some trouble parsing the use case that isn't handled internally in the provider.

> Some providers may draw a distinction between boolean flag value evaluation, and flag activity. Put another way, a flag may be "disabled", but simultaneously have an associated auxillary boolean payload of `true`.

Author: >This single API will tell you "Is this enabled in the UI" and also "is the flag evaluating to true"?

Not quite (unless I'm misunderstanding you)

The issue is  that some providers allow logically boolean flags (enabled/disabled) WITH separated associated boolean payloads (every flag has a bool and an optional number/string/boolean payload). Because some providers work that way, I think it forces us to have this kind of resolution - we need a method for the boolean flag, and a separate method to get it's associated boolean payload.

In providers that DONT have this concept, `isEnabled()` and `getBooleanValue()` do the same thing

> Some providers may draw a distinction between boolean flag value evaluation, and flag activity. Put another way, a flag may be "disabled", but simultaneously have an associated auxillary boolean payload of `true`.

Author: @InTheCloudDan 

>Is there additional information around this? I'm having some trouble parsing the use case that isn't handled internally in the provider.

See: https://github.com/open-feature/spec/pull/35/files#r855378994. It is really internally handled by the provider, but I wanted to call it out here because it's relevant to the expected behavior of the evaluation API.

Maybe this proposal isn't the best solution, but @dabeeeenster has brought the issue up a few times, and this is the best I can come up with. Open to more suggestions though.

> Some providers may draw a distinction between boolean flag value evaluation, and flag activity. Put another way, a flag may be "disabled", but simultaneously have an associated auxillary boolean payload of `true`.

Reviewer 5: I'm also confused by this but it seems others are on top of clarifications

> Some providers may draw a distinction between boolean flag value evaluation, and flag activity. Put another way, a flag may be "disabled", but simultaneously have an associated auxillary boolean payload of `true`.

Reviewer: @toddbaert Which providers do that? A flag evaluation that's "off" seems like it should return the default value and a reason for it.

> Some providers may draw a distinction between boolean flag value evaluation, and flag activity. Put another way, a flag may be "disabled", but simultaneously have an associated auxillary boolean payload of `true`.

Author: @justinabrahms  Flagsmith, particularly behaves this way. It can return false and a payload.

> Some providers may draw a distinction between boolean flag value evaluation, and flag activity. Put another way, a flag may be "disabled", but simultaneously have an associated auxillary boolean payload of `true`.

Reviewer 6: Just to chime in here - we see the notion of a flag being `false` and consequently disabled or providing some form of convention in some way as something of an anti-pattern. There are use cases where flags are used in a manner that they travel from true > false for good reason - e.g. retiring a feature. 

We've designed Flagsmith in a manner that `false` does not impute anything in that regard - it's just something that is `false` and up to the developer to interpret that however they see fit.

> Some providers may draw a distinction between boolean flag value evaluation, and flag activity. Put another way, a flag may be "disabled", but simultaneously have an associated auxillary boolean payload of `true`.

Author: So, in Flagsmith, I can setup a single flag, let's call it `toddsFlag`. Like all flags in Flagsmith, it has a boolean and an optional `boolean | string | number` auxiliary value, depending on how you've configuration this it in this case. Let's consider the auxiliary flag is boolean for the rest of this discussion.

This same flag can used by the flagsmith SDK 2 ways:

`flagsmith.hasFeature('toddsFlag')` which returns a boolean corresponding to the big toggle switch on the Flagsmith UI. 

AND 

`flagsmith.getValue('toddsFlag')` which can return a boolean completely independent of the returned value of `hasFeature` above, true or false.

It was supporting this that I was attempting to do, but after this discussion, I'm wondering if it's not better to completely remove the `isEnabled()` method entirely, and let the Flagsmith provider determine how to handle resolving `getBooleanValue` in their case.

> Some providers may draw a distinction between boolean flag value evaluation, and flag activity. Put another way, a flag may be "disabled", but simultaneously have an associated auxillary boolean payload of `true`.

Reviewer 5: Hmm thanks for the clarification. The wording here makes more sense with this in mind.

> It was supporting this that I was attempting to do, **but after this discussion, I'm wondering if it's not better to completely remove the `isEnabled()` method entirely**, and let the Flagsmith provider determine how to handle resolving `getBooleanValue` in their case.

I'm supportive of this tbh. Seems very provider-specific

> Some providers may draw a distinction between boolean flag value evaluation, and flag activity. Put another way, a flag may be "disabled", but simultaneously have an associated auxillary boolean payload of `true`.

Reviewer 7: I'm in favor of removing `isEnabled()`. In my opinion, it seems very unlikely that an application author would expect `isEnabled()` and `getBooleanValue()` to behave differently. Removing it would eliminate that confusion.

> Some providers may draw a distinction between boolean flag value evaluation, and flag activity. Put another way, a flag may be "disabled", but simultaneously have an associated auxillary boolean payload of `true`.

Reviewer 6: > flagsmith.getValue('toddsFlag') which can return a boolean completely independent of the returned value of hasFeature above, true or false. @dabeeeenster please correct me here, if I'm wrong.

Correct. 

This is something of a hangover from pure feature flags and a more complete flag and remote config solution. We wanted to give people a simple boolean state for a flag but then be able to enhance it with essentially a general String value, but then what if they store `True` in the string value, and how does that typing get supported through all the SDKs and their various type handling.

Hard to say what the best general solution is here. I think if the OpenFeature SDK can provide its best interpretation of the intention from the provider's interface but let developers have access to the base data is probably the optimal solution. 

I believe most vendors have a similar approach of booleans and additional config/multi-variate values but I cant (and don't want to !) speak for them.

> Some providers may draw a distinction between boolean flag value evaluation, and flag activity. Put another way, a flag may be "disabled", but simultaneously have an associated auxillary boolean payload of `true`.

Reviewer 7: If this functionality is there due to historical reasons, I propose that we do not add this to the spec. I think the expected behavior for OpenFeature application developers would be that a disabled feature flag does not have any meaningful auxiliary value. In this case, it would simply return the default value provided during the evaluation. However, the provider pattern is flexible enough to allow each vendor to decide how best to implement the interface. In this case, the `getBooleanValue()` could get the value from the state of the flag or from the `getValue()` method.

> Some providers may draw a distinction between boolean flag value evaluation, and flag activity. Put another way, a flag may be "disabled", but simultaneously have an associated auxillary boolean payload of `true`.

Author: For now, I've removed the `isEnabled()` and `isEnabledDetails()` methods. We can always open up a new issue for this if we think it's something that's absolutely necessary, but I agree that the `provider` interface (spec to come) gives enough flexibility to address this in numerous ways.

##### Requirement 1.10

> > The `client` **MUST** provide methods for flag value evaluation with the value wrapped in a `evaluation details` object, with parameters `flag key` (string, required), `default value` (boolean | number | string | object, required), `evaluation context` (optional), and `evaluation options` (optional).

Reviewer 5: You use language-leaning wording again here, but it is less egregious so i'll let you decide if you want to change it.

Author: I will change `object` to `interface`.

```
// example detailed boolean flag evaluation
FlagEvaluationDetails<boolean> myBoolDetails = client.getBooleanDetails('bool-flag', false);

// example detailed string flag evaluation
FlagEvaluationDetails<string> myStringDetails = client.getStringDetails('string-flag', 'N/A', evaluationContext, options);

```

##### Requirement 1.11

> > The `evaluation details` wrapper **MUST** have a `value` field matching the type of the flag value, and containing to the flag value.

Reviewer 5: ```suggestion
> The `evaluation details` wrapper **MUST** have a `value` field matching the type of the flag value, and containing the flag value.
```

What does "matching the type of the flag value" mean?

##### Condition 1.12

> The language supports generics (or an equivalent feature).
>

> > ##### Conditional Requirement 1.12.1

Author: Ah, I see that. It didn't happen on my local preview. I think the spec parser might require this, but I can tweak it. It looks somewhat ugly this way.

>

> > > The `evaluation details` wrapper **SHOULD** accept a generic argument (or utilize an equivalent language feature) which indicates the type of the wrapped `value` field.

Reviewer 5: What happens if there is a type mismatch?

Author: I was going to mention more about this in this provider spec, but I should stay in this spec that all the flag evaluation methods should default when there's a type mismatch (for the provider spec, the provider should throw or otherwise indicate abnormal execution).

##### Requirement 1.13

> > The `evaluation details` wrapper **MUST** have a `flag key` field field of type string, representing the flag key.

Reviewer 5: ```suggestion
> The `evaluation details` wrapper **MUST** have a `flag key` field of type string, representing the flag key.
```

##### Requirement 1.14

> > The `evaluation details` wrapper **MUST** have an optional `variant` field of type string which represents a space-constrained identifier for a particular flag value.

Reviewer 5: What do you mean space-constrained?

##### Requirement 1.15

> > The `evaluation details` wrapper **MUST** have a `reason` field of type string indicating the semantic reason for the returned flag value. Possible values vary by provider, but may include `"TARGETING_MATCH"`, `"SPLIT"`, `"DISABLED"`, `"DEFAULT"`, `"UNKNOWN"` and `"ERROR"`.

Reviewer 5: Seems like this could be an optional field

Author: Agreed.

##### Requirement 1.16

> In cases of abnormal execution (network failure, unhandled error, etc) the `reason` field in the `evaluation details` **MUST** be set to `"ERROR"`.

##### Requirement 1.17

> > The `evaluation details` wrapper **MUST** have an optional `error code` field of type string, which, in cases of abnormal execution, identifies an error occurred during flag evaluation, with possible values `"PROVIDER_NOT_READY"`, `"FLAG_NOT_FOUND"`, `"PARSE_ERROR"`, `"TYPE_MISMATCH"`, or `"GENERAL"`.

Author: It all goes back to https://github.com/open-feature/spec/issues/32#issuecomment-1101612389 and the linked discussion. I think we should decide if we want this to be free-form or not. This is an open question as I mentioned here: https://github.com/open-feature/spec/pull/35/files/a782d675f1aef6924092f7b9421a53c2f2ce449f#r855313094

> > The `evaluation details` wrapper **MUST** have an optional `error code` field of type string, which, in cases of abnormal execution, identifies an error occurred during flag evaluation, with possible values `"PROVIDER_NOT_READY"`, `"FLAG_NOT_FOUND"`, `"PARSE_ERROR"`, `"TYPE_MISMATCH"`, or `"GENERAL"`.

Author: @justinabrahms @InTheCloudDan @beeme1mr  How do you think we should proceed here? At this point, I dont mind which path we go down (enum or string) at this point, I'm sure if we make the wrong choice it will become apparent in alpha. Please take a look at the linked discussion. We can also take a vote if that's preferable.

> > The `evaluation details` wrapper **MUST** have an optional `error code` field of type string, which, in cases of abnormal execution, identifies an error occurred during flag evaluation, with possible values `"PROVIDER_NOT_READY"`, `"FLAG_NOT_FOUND"`, `"PARSE_ERROR"`, `"TYPE_MISMATCH"`, or `"GENERAL"`.

Reviewer: No strong opinions yet. Happy to do it in strings in javascript and enums in java and see which ones we like. :)

> > The `evaluation details` wrapper **MUST** have an optional `error code` field of type string, which, in cases of abnormal execution, identifies an error occurred during flag evaluation, with possible values `"PROVIDER_NOT_READY"`, `"FLAG_NOT_FOUND"`, `"PARSE_ERROR"`, `"TYPE_MISMATCH"`, or `"GENERAL"`.

Author: OK. I think we'll keep the related discussion open so we remember to come back to it, and I'll go forward with strings for now in the spec. I'm totally open to discuss it more, nothing here is set in stone.

##### Requirement 1.18

> > If the `reason` in the `evaluation details` is set to `"ERROR"`, the sibling `error code` field **MUST** be set.

Author: This is dependent on the outcome of the discussion here: https://github.com/open-feature/spec/issues/32#issuecomment-1101612389

Reviewer 5: Error code was quite a contentious topic in opentelemetry which we went back and forth on for a long time. I would think about what usecase you're trying to solve by having an error code and see if you can solve it in a more explicit way. Maybe a "retryable" boolean?

Reviewer 7: The main use cases we're focusing on is telemetry and debugging.

##### Requirement 1.19

> The `evaluation details` wrapper **MUST** have a `executed hooks` field, containing an object which itself contains 4 arrays, identified with keys (`before`, `after`, `error` and `finally`) indicating the hooks run for the respective life-cycle stages.

```
executedHooks: {
  before: [],
  after: [],
  error: [],

>   finally: []

Reviewer: Not sure how we want to handle it, but 'finally' is a java reserved word. I've chosen "afterAll" instead.

Author: Oh, right. I like `finally` because really these sort of hooks behave just like a `finally` block, so it's intuitive. But if it's a reserved word in Java, it may be in other languages as well (frankly I'm a bit surprised the TS parser has no issue with it).

I'm OK with making it "afterAll" in general, or we could say it's `finally` OR `afterAll`.

Reviewer 5: Avoiding using well known reserved words is a good idea, but in general it may be impossible to completely avoid reserved words. Maybe add some guidance to the spec that if a reserved word is used then choose an appropriate substitute.

Author: We're going to remove this for now anyway, as discussed in other comments, unless anyone disagrees. @justinabrahms @dyladan

}
```

##### Requirement 1.20

> > The value of the elements in the arrays contained in the `executed hooks` object **SHOULD** allow authors to uniquely identify the hooks executed during the flag evaluation lifecycle. This could be done with a mandatory string field defined on the `hook` interface, or by using the fully qualified class name in languages where such concepts exist.

Reviewer 5: This seems like a feature that I would not include in version 1. If there is need for it then it can always be added later, but for now it seems like trying to cover all the bases before you know if you are even playing baseball.

> > The value of the elements in the arrays contained in the `executed hooks` object **SHOULD** allow authors to uniquely identify the hooks executed during the flag evaluation lifecycle. This could be done with a mandatory string field defined on the `hook` interface, or by using the fully qualified class name in languages where such concepts exist.

Author: @beeme1mr I guess I agree with Dan's point here.

> > The value of the elements in the arrays contained in the `executed hooks` object **SHOULD** allow authors to uniquely identify the hooks executed during the flag evaluation lifecycle. This could be done with a mandatory string field defined on the `hook` interface, or by using the fully qualified class name in languages where such concepts exist.

Reviewer 7: It doesn't have to be in the spec but it was a valuable feature in our experimental project. It allows you to understand what hooks ran and in what order.

> > The value of the elements in the arrays contained in the `executed hooks` object **SHOULD** allow authors to uniquely identify the hooks executed during the flag evaluation lifecycle. This could be done with a mandatory string field defined on the `hook` interface, or by using the fully qualified class name in languages where such concepts exist.

Reviewer 7: After some more thought, I agree. It's a good idea but not necessary in the initial spec.

##### Requirement 1.21

> The `evaluation options` object **MUST** have a `hooks` field which can contain hooks to be executed for the respective flag evaluation.

see [hooks](./hooks.md)

##### Requirement 1.22

> > No methods, functions, or operations on the client should ever throw exceptions, or otherwise abnormally end. Flag evaluation calls must always return the `default value` in the event of abnormal execution.

Reviewer 5: I would carve out an exception for configuration and setup. If something is incorrectly setup, then failing on startup is a reasonable choice and swallowing that error might be confusing.

> > No methods, functions, or operations on the client should ever throw exceptions, or otherwise abnormally end. Flag evaluation calls must always return the `default value` in the event of abnormal execution.

Author: I specifically mentioned "client" here, because most configuration is done by the point you have an instance of one, with the exception of adding hooks to the client... I think it's a fair point though, to say this rule really applies to flag evaluation only.

> > No methods, functions, or operations on the client should ever throw exceptions, or otherwise abnormally end. Flag evaluation calls must always return the `default value` in the event of abnormal execution.

Reviewer: One thing came up for me:  This is still true for the case where something in the `finally()` hook errors out, right? If that's the case, I think it will make the logic quite odd for handling those exceptions, making it not much different than `after`

> > No methods, functions, or operations on the client should ever throw exceptions, or otherwise abnormally end. Flag evaluation calls must always return the `default value` in the event of abnormal execution.

Author: I think the `finally` or `afterAll` hook is pretty similar functionally to the `after` hook. In my mind, the big difference is that the `after` hook would only be executed if flag evaluation proceeds normally, while the `finally` hook executes in both normal, and abnormal `error` execution. What do you think?

##### Condition 1.23

> The language type system differentiates between floating and integer type numeric values.
>
> ##### Conditional Requirement 1.23.1
>

> > > The `client` **MUST** provide functions for the resolution of both floating point and integer numeric values.

Reviewer: Do you see this as `getFloatDetails` and `getIntegerDetails` instead of `getNumberDetails`?

Author: > Do you see this as getFloatDetails and getIntegerDetails instead of getNumberDetails?

That's pretty much what I had in mind, ya. Cloudbees Feature Management has something similar.

Reviewer 5: Maybe an example would help. Also, there could be a language which differentiates between these types but allows for unified handling of them (like python) or languages with more number types than just these. I would leave it up to the language authors to do whatever is idiomatic in their language.

Author: I can change it to a "SHOULD" pointing out that implementors should simply consider this.

##### Condition 1.24

> The language features async/await semantics.
>
> ##### Conditional Requirement 1.24.1
>

> > > The `client` **SHOULD** provide asynchronous functions for flag evaluation.

Reviewer 5: This is extremely specific. I would generalize it to something like "The client MUST provide a function for flag evaluation which does not block the thread of execution". I would also not gate this on async/await specifically because that is only one of many ways to achieve this goal.

Author: Largely agreed, but that would be changing a "SHOULD" to a "MUST". Is that what you recommend? In multithreaded environments like Java, blocking probably isn't that uncommon.

Reviewer 5: Whether you want SHOULD or MUST, the important part is not the async/await syntax but the execution blocking
